concrete/controllers/dialog/type/update_from_type.php

```diff
@@ -0,0 +1,284 @@
+        }
+
+        if(count($pageTypeDefaultPageBlocksClone) > 0){
+          foreach ($pageTypeDefaultPageBlocksClone as $pageTypeBlock) {
+            $blocksToAdd[] = array('bID' => $pageTypeBlock->getBlockID(), 'actualDisplayOrder' => $pageTypeBlock->cbDisplayOrder, 'pageTypeArHandle' => $pageTypeBlock->getAreaHandle());
+          }
+        }
+
+        $records[] = array(
+          'blocksToAdd' => $blocksToAdd,
+          'blocksToUpdate' => $blocksToUpdate,
+          'cID' => $pagePlusCV['cID'],
+          'cvID' => $pagePlusCV['cvID']
+        );
+
+      }
+
+      $queueName = $queue->getName();
+      $queue->deleteQueue();
+      $queue = Queue::get($queueName);
+
+      foreach ($records as $record) {
+          $queue->send(serialize($record));
+      }
+
+      return $queue;
+    }
+
+    public function submit($ptID, $pTemplateID)
+    {
+
+        $responseFactory = $this->app->make(ResponseFactory::class);
+
+        if (!$this->validateAction() || !$this->canAccess()) {
+          $this->app->shutdown();
+          return;
+        }
+
+        $this->fetchTypeAndTemplate($ptID, $pTemplateID);
+        $pageTypeDefaultPage = $this->pagetype->getPageTypePageTemplateDefaultPageObject($this->template);
+
+        if (!$pageTypeDefaultPage->isMasterCollection()) {
+          $this->app->shutdown();
+          return;
+        }
+
+        $queueName = sprintf('update_pagetype_defaults_%s', $this->pagetype->getPageTypeID());
+        $queue = \Queue::get($queueName);
+
+        if ($_POST['process']) {
+          $db = \Database::connection();
+          $obj = new \stdClass();
+          $messages = $queue->receive(20);
+
+          foreach ($messages as $key => $message) {
+            $record = unserialize($message->body);
+            $childPage = \Page::getByID($record['cID'], $record['cvID']);
+
+            $blocksToUpdate = $record['blocksToUpdate'];
+            $blocksToAdd = $record['blocksToAdd'];
+            $handlesToOrder = array();
+
+            foreach ($blocksToAdd as $blockToAdd) {
+              $pageTypeBlock = \Block::getByID($blockToAdd['bID'], $pageTypeDefaultPage, $blockToAdd['pageTypeArHandle']);
+              $pageTypeBlock->alias($childPage);
+              $addedChildPageblock = \Block::getByID($pageTypeBlock->getBlockID(), $childPage, $blockToAdd['pageTypeArHandle']);
+              $addedChildPageblock->setAbsoluteBlockDisplayOrder($blockToAdd['actualDisplayOrder']);
+            }
+
+            foreach ($blocksToUpdate as $blockToUpdate) {
+              $childPageBlock = \Block::getByID($blockToUpdate['bID'], $childPage, $blockToUpdate['arHandle']);
+              $childPageBlockPerms = new \Permissions($childPageBlock);
+
+              if (!is_object($childPageBlock) || $childPageBlock->isError() || !$childPageBlockPerms->canAdminBlock()) {
+                continue;
+              }
+
+              $actions = $blockToUpdate['actions'];
+
+              foreach ($actions as $action) {
```

Comment:
Not sure where I mentioned to 'serialize a class'? I meant that you have a lot of logic here that's is completely undocumented and hard to follow. It would make perfect sense to extract this big chunk of code to methods and optionally add comments. That would make it easier for other devs to read the code, and e.g. make changes to it.